assets/model_monitoring/components/src/model_data_collector_preprocessor/mdc_preprocessor_helper.py

```diff
@@ -16,23 +18,152 @@
+                if store_url.is_folder_exists(f"{m.strftime('%Y/%m')}") else []
+        return same_month(m.replace(day=1, hour=0, minute=0, second=0), m)
+
+    def end_of_month(m) -> List[str]:
+        if is_start_of_month(m):
+            return [store_url.get_abfs_url(f"{m.strftime('%Y/%m')}/*/*/*.jsonl")] \
+                if store_url.is_folder_exists(f"{m.strftime('%Y/%m')}") else []
+        _, month_days = calendar.monthrange(m.year, m.month)
+        return same_month(m, m.replace(day=month_days, hour=23, minute=59, second=59))
+
+    def start_of_day(d) -> List[str]:
+        if is_end_of_day(d):
+            return [store_url.get_abfs_url(f"{d.strftime('%Y/%m/%d')}/*/*.jsonl")] \
+                if store_url.is_folder_exists(f"{d.strftime('%Y/%m/%d')}") else []
+        return same_day(d.replace(hour=0, minute=0, second=0), d)
+
+    def end_of_day(d) -> List[str]:
+        if is_start_of_day(d):
+            return [store_url.get_abfs_url(f"{d.strftime('%Y/%m/%d')}/*/*.jsonl")] \
+                if store_url.is_folder_exists(f"{d.strftime('%Y/%m/%d')}") else []
+        return same_day(d, d.replace(hour=23, minute=59, second=59))
+
+    def cross_year(start, end) -> List[str]:
+        middle_years = [
+            store_url.get_abfs_url(f"{y}/*/*/*/*.jsonl") for y in range(start.year+1, end.year)
+            if store_url.is_folder_exists(f"{y}")
+        ]
+        return end_of_year(start) + middle_years + start_of_year(end)
+
+    def cross_month(start: datetime, end) -> List[str]:
+        _start = (start + relativedelta(months=1)).replace(day=1, hour=1)
+        _end = end.replace(day=1, hour=0)  # skip last month
+        middle_months = [
+            store_url.get_abfs_url(f"{m.strftime('%Y/%m')}/*/*/*.jsonl")
+            for m in date_range(_start, _end, relativedelta(months=1))
+            if store_url.is_folder_exists(f"{m.strftime('%Y/%m')}")
+        ]
+        return end_of_month(start) + middle_months + start_of_month(end)
+
+    def cross_day(start, end) -> List[str]:
+        _start = (start + timedelta(days=1)).replace(hour=1)
+        _end = end.replace(hour=0)  # skip last day
+        middle_days = [
+            store_url.get_abfs_url(f"{d.strftime('%Y/%m/%d')}/*/*.jsonl")
+            for d in date_range(_start, _end, timedelta(days=1))
+            if store_url.is_folder_exists(f"{d.strftime('%Y/%m/%d')}")
+        ]
+        return end_of_day(start) + middle_days + start_of_day(end)
+
+    def cross_hour(start, end) -> List[str]:
+        _start = start.replace(minute=0)
+        _end = end.replace(minute=59)
+        return [
+            store_url.get_abfs_url(f"{h.strftime('%Y/%m/%d/%H')}/*.jsonl")
+            for h in date_range(_start, _end, timedelta(hours=1))
+            if store_url.is_folder_exists(f"{h.strftime('%Y/%m/%d/%H')}")
+        ]
+
+    # ensure start_datetime and end_datetime both or neither have tzinfo
```

Comment:
What does this mean?

Reply:
python cannot do datetime comparing if one have tzinfo the other has not.